File: android/app/src/main/java/com/ryanladuca/myafbase/data/prefs/PFRAProfileStore.kt
package com.ryanladuca.myafbase.data.prefs

import android.content.Context
import android.content.SharedPreferences
import com.ryanladuca.myafbase.domain.logic.PFRACardioEvent
import com.ryanladuca.myafbase.domain.logic.PFRACoreEvent
import com.ryanladuca.myafbase.domain.logic.PFRAGender
import com.ryanladuca.myafbase.domain.logic.PFRAStrengthEvent
import com.ryanladuca.myafbase.domain.logic.PFRATargetTier
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class PFRAProfile(
    val gender: PFRAGender = PFRAGender.MALE,
    val age: Int = 28,
    val heightFeet: Int = 5,
    val heightInches: Int = 10,
    val waistTenths: Int = 340,
    val cardioEvent: PFRACardioEvent = PFRACardioEvent.TWO_MILE_RUN,
    val runMinutes: Int = 15,
    val runSeconds: Int = 0,
    val hamrShuttles: Int = 60,
    val strengthEvent: PFRAStrengthEvent = PFRAStrengthEvent.PUSH_UPS,
    val strengthReps: Int = 40,
    val coreEvent: PFRACoreEvent = PFRACoreEvent.SIT_UPS,
    val coreReps: Int = 45,
    val plankMinutes: Int = 2,
    val plankSeconds: Int = 0,
    val targetTier: PFRATargetTier = PFRATargetTier.SATISFACTORY
) {
    val heightTotalInches: Double get() = (heightFeet * 12 + heightInches).toDouble()
    val waistInches: Double get() = waistTenths / 10.0
    val cardioValue: Double
        get() = when (cardioEvent) {
            PFRACardioEvent.TWO_MILE_RUN -> (runMinutes * 60 + runSeconds).toDouble()
            PFRACardioEvent.HAMR -> hamrShuttles.toDouble()
        }
    val coreValue: Double
        get() = when (coreEvent) {
            PFRACoreEvent.FOREARM_PLANK -> (plankMinutes * 60 + plankSeconds).toDouble()
            else -> coreReps.toDouble()
        }
}

class PFRAProfileStore(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("pfra_profile", Context.MODE_PRIVATE)

    private val _profile = MutableStateFlow(load())
    val profile: StateFlow<PFRAProfile> = _profile.asStateFlow()

    fun update(transform: (PFRAProfile) -> PFRAProfile) {
        val next = transform(_profile.value)
        _profile.value = next
        save(next)
    }

    private fun load(): PFRAProfile = PFRAProfile(
        gender = runCatching {
            PFRAGender.valueOf(prefs.getString("gender", PFRAGender.MALE.name)!!)
        }.getOrDefault(PFRAGender.MALE),
        age = prefs.getInt("age", 28),
        heightFeet = prefs.getInt("heightFeet", 5),
        heightInches = prefs.getInt("heightInches", 10),
        waistTenths = prefs.getInt("waistTenths", 340),
        cardioEvent = runCatching {
            PFRACardioEvent.valueOf(prefs.getString("cardioEvent", PFRACardioEvent.TWO_MILE_RUN.name)!!)
        }.getOrDefault(PFRACardioEvent.TWO_MILE_RUN),
        runMinutes = prefs.getInt("runMinutes", 15),
        runSeconds = prefs.getInt("runSeconds", 0),
        hamrShuttles = prefs.getInt("hamrShuttles", 60),
        strengthEvent = runCatching {
            PFRAStrengthEvent.valueOf(prefs.getString("strengthEvent", PFRAStrengthEvent.PUSH_UPS.name)!!)
        }.getOrDefault(PFRAStrengthEvent.PUSH_UPS),
        strengthReps = prefs.getInt("strengthReps", 40),
        coreEvent = runCatching {
            PFRACoreEvent.valueOf(prefs.getString("coreEvent", PFRACoreEvent.SIT_UPS.name)!!)
        }.getOrDefault(PFRACoreEvent.SIT_UPS),
        coreReps = prefs.getInt("coreReps", 45),
        plankMinutes = prefs.getInt("plankMinutes", 2),
        plankSeconds = prefs.getInt("plankSeconds", 0),
        targetTier = runCatching {
            PFRATargetTier.valueOf(prefs.getString("targetTier", PFRATargetTier.SATISFACTORY.name)!!)
        }.getOrDefault(PFRATargetTier.SATISFACTORY)
    )

    private fun save(p: PFRAProfile) {
        prefs.edit()
            .putString("gender", p.gender.name)
            .putInt("age", p.age)
            .putInt("heightFeet", p.heightFeet)
            .putInt("heightInches", p.heightInches)
            .putInt("waistTenths", p.waistTenths)
            .putString("cardioEvent", p.cardioEvent.name)
            .putInt("runMinutes", p.runMinutes)
            .putInt("runSeconds", p.runSeconds)
            .putInt("hamrShuttles", p.hamrShuttles)
            .putString("strengthEvent", p.strengthEvent.name)
            .putInt("strengthReps", p.strengthReps)
            .putString("coreEvent", p.coreEvent.name)
            .putInt("coreReps", p.coreReps)
            .putInt("plankMinutes", p.plankMinutes)
            .putInt("plankSeconds", p.plankSeconds)
            .putString("targetTier", p.targetTier.name)
            .apply()
    }
}

File: MyAFBase/MyAFBase/Services/PFRAProfileStore.swift
import Foundation
import Observation

/// Shared PFRA inputs for Score Calculator and Goal Planner.
/// Persisted in `UserDefaults` so edits in one tool carry into the other.
@Observable
@MainActor
final class PFRAProfileStore {
    private enum Keys {
        static let gender = "pfraProfile.gender"
        static let age = "pfraProfile.age"
        static let heightFeet = "pfraProfile.heightFeet"
        static let heightInches = "pfraProfile.heightInches"
        static let waistTenths = "pfraProfile.waistTenths"
        static let cardioEvent = "pfraProfile.cardioEvent"
        static let runMinutes = "pfraProfile.runMinutes"
        static let runSeconds = "pfraProfile.runSeconds"
        static let hamrShuttles = "pfraProfile.hamrShuttles"
        static let strengthEvent = "pfraProfile.strengthEvent"
        static let strengthReps = "pfraProfile.strengthReps"
        static let coreEvent = "pfraProfile.coreEvent"
        static let coreReps = "pfraProfile.coreReps"
        static let plankMinutes = "pfraProfile.plankMinutes"
        static let plankSeconds = "pfraProfile.plankSeconds"
        static let targetTier = "pfraProfile.targetTier"
    }

    var gender: PFRAGender {
        didSet { defaults.set(gender.rawValue, forKey: Keys.gender) }
    }

    var age: Int {
        didSet { defaults.set(age, forKey: Keys.age) }
    }

    var heightFeet: Int {
        didSet { defaults.set(heightFeet, forKey: Keys.heightFeet) }
    }

    var heightInches: Int {
        didSet { defaults.set(heightInches, forKey: Keys.heightInches) }
    }

    var waistTenths: Int {
        didSet { defaults.set(waistTenths, forKey: Keys.waistTenths) }
    }

    var cardioEvent: PFRACardioEvent {
        didSet { defaults.set(cardioEvent.rawValue, forKey: Keys.cardioEvent) }
    }

    var runMinutes: Int {
        didSet { defaults.set(runMinutes, forKey: Keys.runMinutes) }
    }

    var runSeconds: Int {
        didSet { defaults.set(runSeconds, forKey: Keys.runSeconds) }
    }

    var hamrShuttles: Int {
        didSet { defaults.set(hamrShuttles, forKey: Keys.hamrShuttles) }
    }

    var strengthEvent: PFRAStrengthEvent {
        didSet { defaults.set(strengthEvent.rawValue, forKey: Keys.strengthEvent) }
    }

    var strengthReps: Int {
        didSet { defaults.set(strengthReps, forKey: Keys.strengthReps) }
    }

    var coreEvent: PFRACoreEvent {
        didSet { defaults.set(coreEvent.rawValue, forKey: Keys.coreEvent) }
    }

    var coreReps: Int {
        didSet { defaults.set(coreReps, forKey: Keys.coreReps) }
    }

    var plankMinutes: Int {
        didSet { defaults.set(plankMinutes, forKey: Keys.plankMinutes) }
    }

    var plankSeconds: Int {
        didSet { defaults.set(plankSeconds, forKey: Keys.plankSeconds) }
    }

    var targetTier: PFRATargetTier {
        didSet { defaults.set(targetTier.rawValue, forKey: Keys.targetTier) }
    }

    var heightTotalInches: Double {
        Double(heightFeet * 12 + heightInches)
    }

    var waistInches: Double {
        Double(waistTenths) / 10.0
    }

    var cardioValue: Double {
        switch cardioEvent {
        case .twoMileRun:
            Double(runMinutes * 60 + runSeconds)
        case .hamr:
            Double(hamrShuttles)
        }
    }

    var coreValue: Double {
        switch coreEvent {
        case .sitUps, .crossLegReverseCrunch:
            Double(coreReps)
        case .forearmPlank:
            Double(plankMinutes * 60 + plankSeconds)
        }
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults

        gender = defaults.string(forKey: Keys.gender).flatMap(PFRAGender.init(rawValue:)) ?? .male
        age = Self.int(defaults, Keys.age, default: 25)
        heightFeet = Self.int(defaults, Keys.heightFeet, default: 5)
        heightInches = Self.int(defaults, Keys.heightInches, default: 9)
        waistTenths = Self.int(defaults, Keys.waistTenths, default: 320)
        cardioEvent = defaults.string(forKey: Keys.cardioEvent).flatMap(PFRACardioEvent.init(rawValue:)) ?? .twoMileRun
        runMinutes = Self.int(defaults, Keys.runMinutes, default: 15)
        runSeconds = Self.int(defaults, Keys.runSeconds, default: 0)
        hamrShuttles = Self.int(defaults, Keys.hamrShuttles, default: 50)
        strengthEvent = defaults.string(forKey: Keys.strengthEvent).flatMap(PFRAStrengthEvent.init(rawValue:)) ?? .pushUps
        strengthReps = Self.int(defaults, Keys.strengthReps, default: 40)
        coreEvent = defaults.string(forKey: Keys.coreEvent).flatMap(PFRACoreEvent.init(rawValue:)) ?? .sitUps
        coreReps = Self.int(defaults, Keys.coreReps, default: 45)
        plankMinutes = Self.int(defaults, Keys.plankMinutes, default: 2)
        plankSeconds = Self.int(defaults, Keys.plankSeconds, default: 0)
        targetTier = defaults.string(forKey: Keys.targetTier).flatMap(PFRATargetTier.init(rawValue:)) ?? .satisfactory
    }

    func clampHeightInches() {
        if heightFeet == 7 {
            heightInches = min(heightInches, 11)
        }
    }

    private static func int(_ defaults: UserDefaults, _ key: String, default fallback: Int) -> Int {
        guard defaults.object(forKey: key) != nil else { return fallback }
        return defaults.integer(forKey: key)
    }
}
